### PR TITLE
API/UCP/UCT/IODEMO: Add functions to query the ucp and uct endpoints

### PR DESCRIPTION
**Pull request by leibin2014**
## What
Add functions to query the ucp and uct endpoints

## Why ?
Sometimes users need API to get remote and local IP/port per endpoint

## How ?
Get the remote and local IP/port through RDMA_CM

### CODE REVIEW

**Comment by leibin2014** [comment]
bot:pipe:retest

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -4518,17 +4520,29 @@ typedef struct ucp_ep_attr {
      * this name.
      */
     char     name[UCP_ENTITY_NAME_MAX];
+
+    /**
+     * The local sockaddr that this ep is using.
```
Comment (yosefe) [inline src/ucp/api/ucp.h:4525]: Local socket address for this endpoint. Valid only for endpoints created by connecting to a socket address.

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -4518,17 +4520,29 @@ typedef struct ucp_ep_attr {
...
+     */
+    struct sockaddr_storage local_sockaddr;
+
+    /**
+     * The remote sockaddr this ep is connected to.
```
Comment (yosefe) [inline src/ucp/api/ucp.h:4530]: Remote socket address this endpoint is connected to. Valid only for endpoints created by connecting to a socket address.

File: src/ucp/api/ucp.h [hunk 3]
```diff
@@ -4518,17 +4520,29 @@ typedef struct ucp_ep_attr {
...
 
 /**
- * @ingroup UCP_ENDPOINT
- * @brief Get attributes specific to a particular endpoint.
+ * @ingroup UCP_WORKER
+ * @brief Get attributes of a given endpoint.
  *
  * This routine fetches information about the endpoint.
+ * It should be called only if the endpoint was created with the
+ * UCP_EP_PARAM_FIELD_SOCK_ADDR or UCP_EP_PARAM_FIELD_CONN_REQUEST flags.
```
Comment (yosefe) [inline src/ucp/api/ucp.h:4542]: pls remove since some of endpoint attributes are not related to sockaddr

File: src/ucp/core/ucp_ep.c [hunk 4]
```diff
@@ -2958,5 +2963,41 @@ ucs_status_t ucp_ep_query(ucp_ep_h ep, ucp_ep_attr_t *attr)
...
+
+    if (uct_cm_ep_attr.field_mask & UCT_EP_ATTR_FIELD_LOCAL_SOCKADDR) {
+        ucp_sockaddr_copy_always((struct sockaddr *)&attr->local_sockaddr,
+                                 (struct sockaddr *)&uct_cm_ep_attr.local_address);
+    }
+    
+    if (uct_cm_ep_attr.field_mask & UCT_EP_ATTR_FIELD_REMOTE_SOCKADDR) {
+        ucp_sockaddr_copy_always((struct sockaddr *)&attr->remote_sockaddr,
+                                 (struct sockaddr *)&uct_cm_ep_attr.remote_address);
+    }
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:2999]: let's put the sockaddr query code in a separate function:
```
...
if (attr->field_mask & (LOCAL_SOCKADDR|REMOTE_SOCKADDR)) {
   status = ucp_ep_query_sockaddr(ep, attr);
   if (status != UCS_OK) {
     return status;
   }
}

return UCS_OK;
```

File: src/ucs/sys/sock.c [hunk 5]
```diff
@@ -850,6 +850,19 @@ ucs_status_t ucs_sockaddr_copy(struct sockaddr *dst_addr,
...
+                              const struct sockaddr *src_addr)
+{
+    ucs_status_t status;
+
+    if (!ucs_sockaddr_is_known_af(src_addr)) {
+        memset(dst_addr, 0, sizeof(*dst_addr));
+    } else {
+        status = ucs_sockaddr_copy(dst_addr, src_addr);
+        ucs_assert_always(status == UCS_OK);
+    }
```
Comment (yosefe) [inline src/ucs/sys/sock.c:863]: if AF not known ucs_sockaddr_copy would fail anyway

File: src/uct/api/tl.h [hunk 6]
```diff
@@ -344,6 +346,7 @@ typedef struct uct_iface_ops {
     uct_ep_connect_func_t               ep_connect;
     uct_ep_disconnect_func_t            ep_disconnect;
     uct_cm_ep_conn_notify_func_t        cm_ep_conn_notify;
+    uct_ep_query_func_t                 ep_query;
```
Comment (yosefe) [inline src/uct/api/tl.h:349]: pls add this to internal_ops to avoid UCT ABI break

File: src/uct/api/uct.h [hunk 7]
```diff
@@ -764,6 +764,21 @@ enum uct_cm_attr_field {
 };
 
 
+/**
```
Comment (yosefe) [inline src/uct/api/uct.h:767]: pls add the new UCT APIs to uct_v2.h

File: src/uct/tcp/tcp_sockcm.c [hunk 8]
```diff
@@ -152,6 +152,7 @@ static uct_iface_ops_t uct_tcp_sockcm_iface_ops = {
     .ep_check                 = (uct_ep_check_func_t)ucs_empty_function_return_unsupported,
     .ep_create                = (uct_ep_create_func_t)ucs_empty_function_return_unsupported,
     .iface_flush              = (uct_iface_flush_func_t)ucs_empty_function_return_unsupported,
+    .ep_query                 = (uct_ep_query_func_t)ucs_empty_function_return_unsupported,
```
Comment (yosefe) [inline src/uct/tcp/tcp_sockcm.c:155]: TCP should also support query

File: test/apps/iodemo/ucx_wrapper.cc [hunk 9]
```diff
@@ -960,6 +988,7 @@ void UcxConnection::established(ucs_status_t status)
     if (!_use_am && (status == UCS_OK)) {
         assert(_remote_conn_id != 0);
         UCX_CONN_LOG << "Remote id is " << _remote_conn_id;
+        print_addresses();
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc:991]: can we print the addresses in same line with remote id, to reduce number of log messages?

File: test/apps/iodemo/ucx_wrapper.cc [hunk 10]
```diff
@@ -884,6 +886,32 @@ void UcxConnection::set_log_prefix(const struct sockaddr* saddr,
...
+
+    if (_ep == NULL) {
+        return;
+    }
+
+    ep_attr.field_mask = UCP_EP_ATTR_FIELD_LOCAL_SOCKADDR |
+                         UCP_EP_ATTR_FIELD_REMOTE_SOCKADDR;
+
+    ucs_status_t status = ucp_ep_query(_ep, &ep_attr);
+    if (status == UCS_OK) {
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc:901]:
```
if (status != UCS_OK) {
  UCX_CONN_LOG << "ucp_ep_query() failed: " << ucs_status_string(status);
  return;
}
```

File: test/apps/iodemo/ucx_wrapper.cc [hunk 11]
```diff
@@ -884,6 +886,32 @@ void UcxConnection::set_log_prefix(const struct sockaddr* saddr,
     memcpy(_log_prefix, ss.str().c_str(), length);
 }
 
+void UcxConnection::print_addresses()
+{
+    ucp_ep_attr_t ep_attr;
+
+    if (_ep == NULL) {
+        return;
+    }
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc:895]: why needed?

File: test/apps/iodemo/ucx_wrapper.cc [hunk 12]
```diff
@@ -677,6 +677,8 @@ void UcxConnection::disconnect(UcxCallback *callback)
     assert(_establish_cb == NULL);
     assert(_disconnect_cb == NULL);
 
+    print_addresses();
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc:680]: let's not print addresses on disconnect (IMO it's redundant)

**Review by yosefe — COMMENTED** [review]
pls also add gtest for UCP and UCT query

File: src/uct/tcp/tcp_sockcm.c (hunk 8 above)
Comment (leibin2014) [reply src/uct/tcp/tcp_sockcm.c:155]: @yosefe Do you mean also add ep query implementation for tcp_sockcm in this PR?

Comment (yosefe) [reply src/uct/tcp/tcp_sockcm.c:155]: yes, i think it should be pretty small, and it's also needed to make the gtest pass (the gtest which i asked to add)

File: src/ucs/sys/sock.c (hunk 5 above)
Comment (leibin2014) [reply src/ucs/sys/sock.c:863]: @yosefe Yes, so ucs_sockaddr_copy is called only when AF is known. What we should change here?

File: src/uct/api/tl.h (hunk 6 above)
Comment (leibin2014) [reply src/uct/api/tl.h:349]: @yosefe Do you mean move to? Remove the change here, and add to internal_ops?

File: test/apps/iodemo/ucx_wrapper.cc (hunk 11 above)
Comment (leibin2014) [reply test/apps/iodemo/ucx_wrapper.cc:895]: @yosefe If _ep is NULL, calling of ucp_ep_query(_ep, &ep_attr) will have segment fault.

File: test/apps/iodemo/ucx_wrapper.cc (hunk 9 above)
Comment (leibin2014) [reply test/apps/iodemo/ucx_wrapper.cc:988]: @yosefe We can, but not simple. We want to print address when ucp_ep_query status is UCS_OK, but print failure when status is not UCS_OK. using UCX_CONN_LOG to print log in single line can't support condition check. Or we need add more code to deal with it.
Here we only add one line of address log when connection estabilished, so maybe not so important to reduce it?

File: src/ucs/sys/sock.c (hunk 5 above)
Comment (leibin2014) [reply src/ucs/sys/sock.c:876]: @yosefe Or do you mean no need to check the condition ucs_sockaddr_is_known_af(src_addr), just do ucs_sockaddr_copy() anyway?

File: test/apps/iodemo/ucx_wrapper.cc (hunk 9 above)
Comment (leibin2014) [reply test/apps/iodemo/ucx_wrapper.cc]: @yosefe I fixed it.

File: src/ucp/api/ucp.h (hunk 1 above)
Comment (leibin2014) [reply src/ucp/api/ucp.h]: @yosefe Done

File: src/ucp/api/ucp.h (hunk 2 above)
Comment (leibin2014) [reply src/ucp/api/ucp.h]: @yosefe Done

File: src/ucp/api/ucp.h (hunk 3 above)
Comment (leibin2014) [reply src/ucp/api/ucp.h]: @yosefe  Done

File: src/ucp/core/ucp_ep.c (hunk 4 above)
Comment (leibin2014) [reply src/ucp/core/ucp_ep.c]: @yosefe Done

File: src/uct/tcp/tcp_sockcm.c (hunk 8 above)
Comment (leibin2014) [reply src/uct/tcp/tcp_sockcm.c]: @yosefe Done

File: test/apps/iodemo/ucx_wrapper.cc (hunk 10 above)
Comment (leibin2014) [reply test/apps/iodemo/ucx_wrapper.cc]: @yosefe Done

File: test/apps/iodemo/ucx_wrapper.cc (hunk 12 above)
Comment (leibin2014) [reply test/apps/iodemo/ucx_wrapper.cc]: @yosefe Done

File: src/uct/api/tl.h (hunk 6 above)
Comment (leibin2014) [reply src/uct/api/tl.h]: @yosefe Done

File: src/uct/api/uct.h (hunk 7 above)
Comment (leibin2014) [reply src/uct/api/uct.h]: @yosefe Done

**Comment by leibin2014** [comment]
bot:pipe:retest

File: src/ucp/core/ucp_ep.c [hunk 13]
```diff
@@ -2948,6 +2948,45 @@ void ucp_ep_vfs_init(ucp_ep_h ep)
...
 }
 
+static ucs_status_t ucp_ep_query_sockaddr(ucp_ep_h ep, ucp_ep_attr_t *attr)
+{
+    uct_ep_h uct_cm_ep = ucp_ep_get_cm_uct_ep(ep);
+    uct_ep_attr_t uct_cm_ep_attr;
+    ucs_status_t status;
+
+    if (uct_cm_ep == NULL || ucp_is_uct_ep_failed(uct_cm_ep)) {
+        ucs_debug("uct_cm_ep is NULL");
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:2958]: ep %p: no cm

File: src/ucp/core/ucp_ep.c [hunk 14]
```diff
@@ -2949,26 +2948,13 @@ void ucp_ep_vfs_init(ucp_ep_h ep)
...
-#endif
-    }
-
-    if (!(attr->field_mask & UCP_EP_ATTR_FIELD_LOCAL_SOCKADDR) &&
-        !(attr->field_mask & UCP_EP_ATTR_FIELD_REMOTE_SOCKADDR)) {
-        return UCS_OK;
-    }
-
-    if (uct_cm_ep == NULL) {
+    if (uct_cm_ep == NULL || ucp_is_uct_ep_failed(uct_cm_ep)) {
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c:2957]: use ( )

File: src/ucs/sys/sock.c (hunk 5 above)
Comment (yosefe) [reply src/ucs/sys/sock.c:876]: yes, no need to define ucp_sockaddr_copy_always() function
also it's ucs function so should have been uc**s**_sockaddr_copy_always

File: src/uct/api/tl.h [hunk 15]
```diff
@@ -226,6 +226,8 @@ typedef ucs_status_t (*uct_ep_disconnect_func_t)(uct_ep_h ep, unsigned flags);
 
 typedef ucs_status_t (*uct_cm_ep_conn_notify_func_t)(uct_ep_h ep);
 
+typedef ucs_status_t (*uct_ep_query_func_t)(uct_ep_h ep, uct_ep_attr_t *ep_attr);
```
Comment (yosefe) [inline src/uct/api/tl.h:229]: move to internal header file uct_iface.h

File: src/uct/api/v2/uct_v2.h [hunk 16]
```diff
@@ -223,6 +262,21 @@ uct_iface_estimate_perf(uct_iface_h tl_iface, uct_perf_attr_t *perf_attr);
 ucs_status_t uct_md_mem_dereg_v2(uct_md_h md,
                                  const uct_md_mem_dereg_params_t *params);
 
+
+/**
+ * @ingroup UCT_RESOURCE
+ * @brief Get ep's attributes.
+ *
+ * This routine should be called only if the endpoint was created with the
+ * UCT_EP_PARAM_FIELD_SOCKADDR or UCT_EP_PARAM_FIELD_CONN_REQUEST flags.
```
Comment (yosefe) [inline src/uct/api/v2/uct_v2.h:271]: this is documentation for "local_address" and "remote_address" fields. in the future we will want to extend uct_ep_query to return also non-sockaddr related information

File: src/uct/ib/rdmacm/rdmacm_cm.c [hunk 17]
```diff
@@ -825,6 +825,10 @@ static uct_iface_ops_t uct_rdmacm_cm_iface_ops = {
     .iface_is_reachable       = (uct_iface_is_reachable_func_t)ucs_empty_function_return_zero
 };
 
+static uct_iface_internal_ops_t uct_rdmacm_cm_iface_internal_ops = {
```
Comment (yosefe) [inline src/uct/ib/rdmacm/rdmacm_cm.c:828]: pls set default values for all fields, as in uct_base_iface_internal_ops

File: src/uct/tcp/tcp_sockcm.c [hunk 18]
```diff
@@ -165,15 +186,20 @@ static uct_iface_ops_t uct_tcp_sockcm_iface_ops = {
     .iface_is_reachable       = (uct_iface_is_reachable_func_t)ucs_empty_function_return_zero
 };
 
+static uct_iface_internal_ops_t uct_tcp_sockcm_iface_internal_ops = {
```
Comment (yosefe) [inline src/uct/tcp/tcp_sockcm.c:189]: set values for all fields

File: test/apps/iodemo/ucx_wrapper.cc [hunk 19]
```diff
@@ -957,12 +957,33 @@ void UcxConnection::connect_common(ucp_ep_params_t &ep_params,
...
+        } else {
+            UCX_CONN_LOG << "Remote id is " << _remote_conn_id
+                         << ", endpoint " << _ep << ", local address "
+                         << UcxContext::sockaddr_str(
+                                (const struct sockaddr*)&ep_attr.local_sockaddr,
+                                sizeof(ep_attr.local_sockaddr))
+                         << " remote address "
+                         << UcxContext::sockaddr_str(
+                                (const struct sockaddr*)&ep_attr.remote_sockaddr,
+                                sizeof(ep_attr.remote_sockaddr));
```
Comment (yosefe) [inline test/apps/iodemo/ucx_wrapper.cc:983]: looks ugly, let's add local vars 
std::string local_address = ..
std::string remote_address = ..
UCX_CONN_LOG << ...

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 20]
```diff
@@ -549,6 +549,36 @@ class test_ucp_sockaddr : public ucp_test {
...
+    void ep_query()
+    {
+        ucp_ep_attr_t attr;
+
+        attr.field_mask = UCP_EP_ATTR_FIELD_LOCAL_SOCKADDR |
+                          UCP_EP_ATTR_FIELD_REMOTE_SOCKADDR;
+        ucs_status_t status = ucp_ep_query(receiver().ep(), &attr);
+        ASSERT_UCS_OK(status);
+
+        EXPECT_EQ(true, m_test_addr == attr.local_sockaddr);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_sockaddr.cc:561]: use EXPECT_EQ or EXPECT_TRUE(x == y)

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 21]
```diff
@@ -549,6 +549,36 @@ class test_ucp_sockaddr : public ucp_test {
...
+                              m_test_addr.get_port());
+        EXPECT_EQ(true, m_test_addr == attr.remote_sockaddr);
+
+        memset(&attr, 0, sizeof(attr));
+        attr.field_mask = UCP_EP_ATTR_FIELD_LOCAL_SOCKADDR |
+                          UCP_EP_ATTR_FIELD_REMOTE_SOCKADDR;
+        status = ucp_ep_query(sender().ep(), &attr);
+        ASSERT_UCS_OK(status);
+
+        EXPECT_EQ(true, m_test_addr == attr.remote_sockaddr);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_sockaddr.cc:574]: use EXPECT_EQ or EXPECT_TRUE(x == y)

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 22]
```diff
@@ -549,6 +549,36 @@ class test_ucp_sockaddr : public ucp_test {
...
+
+        /* The ports are expected to be different. Ignore them.*/
+        ucs_sockaddr_set_port((struct sockaddr*)&attr.local_sockaddr,
+                              m_test_addr.get_port());
+        EXPECT_EQ(true, m_test_addr == attr.local_sockaddr);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_sockaddr.cc:579]: use EXPECT_EQ or EXPECT_TRUE(x == y)

File: test/gtest/uct/ib/test_sockaddr.cc [hunk 23]
```diff
@@ -216,6 +217,36 @@ class test_uct_sockaddr : public uct_test {
...
+                          UCT_EP_ATTR_FIELD_REMOTE_SOCKADDR;
+        status = uct_ep_query(m_client->ep(index), &attr);
+        ASSERT_UCS_OK(status);
+
+        /* The ports are expected to be different. Ignore them. */
+        ucs_sockaddr_set_port((struct sockaddr*)&attr.local_address,
+                              m_connect_addr.get_port());
+        EXPECT_EQ(true, m_connect_addr == attr.local_address);
+
+        EXPECT_EQ(true, m_connect_addr == attr.remote_address);
```
Comment (yosefe) [inline test/gtest/uct/ib/test_sockaddr.cc:247]: use EXPECT_EQ or EXPECT_TRUE(x == y)

File: src/ucp/core/ucp_ep.c (hunk 13 above)
Comment (leibin2014) [reply src/ucp/core/ucp_ep.c]: @yosefe Done

File: src/ucp/core/ucp_ep.c (hunk 14 above)
Comment (leibin2014) [reply src/ucp/core/ucp_ep.c]: @yosefe Done

File: src/uct/api/tl.h (hunk 15 above)
Comment (leibin2014) [reply src/uct/api/tl.h]: @yosefe Done

File: src/ucs/sys/sock.c (hunk 5 above)
Comment (leibin2014) [reply src/ucs/sys/sock.c]: @yosefe Done

File: src/uct/api/v2/uct_v2.h (hunk 16 above)
Comment (leibin2014) [reply src/uct/api/v2/uct_v2.h]: @yosefe Fixed

File: src/uct/ib/rdmacm/rdmacm_cm.c (hunk 17 above)
Comment (leibin2014) [reply src/uct/ib/rdmacm/rdmacm_cm.c:828]: @yosefe Done

File: src/uct/tcp/tcp_sockcm.c (hunk 18 above)
Comment (leibin2014) [reply src/uct/tcp/tcp_sockcm.c:189]: @yosefe Done

File: test/apps/iodemo/ucx_wrapper.cc (hunk 19 above)
Comment (leibin2014) [reply test/apps/iodemo/ucx_wrapper.cc]: @yosefe Done

File: test/gtest/ucp/test_ucp_sockaddr.cc (hunk 20 above)
Comment (leibin2014) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: @yosefe Done

File: test/gtest/ucp/test_ucp_sockaddr.cc (hunk 21 above)
Comment (leibin2014) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: @yosefe Done

File: test/gtest/ucp/test_ucp_sockaddr.cc (hunk 22 above)
Comment (leibin2014) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: @yosefe Done

File: test/gtest/uct/ib/test_sockaddr.cc (hunk 23 above)
Comment (leibin2014) [reply test/gtest/uct/ib/test_sockaddr.cc]: @yosefe Done. Thanks!

File: test/gtest/ucp/test_ucp_sockaddr.cc [hunk 24]
```diff
@@ -558,25 +558,25 @@ class test_ucp_sockaddr : public ucp_test {
         ucs_status_t status = ucp_ep_query(receiver().ep(), &attr);
         ASSERT_UCS_OK(status);
 
-        EXPECT_EQ(true, m_test_addr == attr.local_sockaddr);
+        EXPECT_TRUE(m_test_addr == attr.local_sockaddr);
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_sockaddr.cc:561]: is it possible to do `EXPECT_EQ(m_test_addr, attr.local_sockaddr)`?
or need to add `std::iostream& operator<<(std::iostream& os, const ucs::sock_addr_storage& addr)` to make it work?

**Comment by leibin2014** [comment]
bot:pipe:retest

File: test/gtest/ucp/test_ucp_sockaddr.cc (hunk 24 above)
Comment (leibin2014) [reply test/gtest/ucp/test_ucp_sockaddr.cc]: @yosefe Changed to EXPECT_EQ

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@shamisp @tonycurtis can you pls take a look?

File: src/ucs/sys/sock.h [hunk 25]
```diff
@@ -428,6 +428,20 @@ ucs_status_t ucs_sock_ipstr_to_sockaddr(const char *ip_str,
...
 
+/**
+ * Extract the IP address from a given socket fd.
+ *
+ * @param [in]   fd          Socket fd.
+ * @param [out]  sock_addr   IP address.
+ * @param [out]  addr_len    Length of the IP address.
+ *
+ * @return ip_str if the sock_addr has a valid IP address or 'Invalid address'
+ *         otherwise.
```
Comment (brminich) [inline src/ucs/sys/sock.h:439]: minor: it returns `UCS_OK` if sock_addr has a valid IP address

File: src/uct/base/uct_cm.c [hunk 26]
```diff
@@ -282,29 +282,31 @@ static ucs_stats_class_t uct_cm_stats_class = {
...
+    self->iface.internal_ops       = internal_ops;
+    self->iface.worker             = ucs_derived_of(worker, uct_priv_worker_t);
+
+    self->iface.md                 = NULL;
+    self->iface.am->arg            = NULL;
+    self->iface.am->flags          = 0;
+    self->iface.am->cb             = (uct_am_callback_t)ucs_empty_function_return_unsupported;
+    self->iface.am_tracer          = NULL;
+    self->iface.am_tracer_arg      = NULL;
+    self->iface.err_handler        = NULL;
```
Comment (brminich) [inline src/uct/base/uct_cm.c:301]: this whole block can be moved one space left, as it was before

**Review by tonycurtis — APPROVED** [review]

File: src/ucp/api/ucp.h [hunk 27]
```diff
@@ -4518,12 +4520,24 @@ typedef struct ucp_ep_attr {
      * this name.
      */
     char     name[UCP_ENTITY_NAME_MAX];
+
+    /**
+     * Local socket address for this endpoint. Valid only for endpoints created 
+     * by connecting to a socket address.
+     */
+    struct sockaddr_storage local_sockaddr;
```
Comment (shamisp) [inline src/ucp/api/ucp.h:4528]: what happens if socket is not connected to a socket address ?

**Comment by shamisp** [comment]
I'm a bit concerned that we are exposing an API that is only useful for "some" endpoints connected using CM or sockets. Can you please clarify your use case beyond: `Sometimes users need API to get remote and local IP/port per endpoint`. thanks

File: src/ucp/api/ucp.h (hunk 27 above)
Comment (leibin2014) [reply src/ucp/api/ucp.h:4528]: @shamisp  Return UCS_ERR_NOT_CONNECTED

File: src/ucs/sys/sock.h (hunk 25 above)
Comment (leibin2014) [reply src/ucs/sys/sock.h]: @brminich  Fixed.

**Comment by leibin2014** [comment]
> I'm a bit concerned that we are exposing an API that is only useful for "some" endpoints connected using CM or sockets. Can you please clarify your use case beyond: `Sometimes users need API to get remote and local IP/port per endpoint`. thanks

@shamisp We provided this API to Tencent in an separate before. Recently another Chinese customer also needs this API. Their argument is that socket has similar API getsockname. Generally, I agree this API is useful for most of users.

File: src/uct/base/uct_cm.c (hunk 26 above)
Comment (leibin2014) [reply src/uct/base/uct_cm.c]: @brminich Good catch. Thanks!

File: src/ucp/api/ucp.h (hunk 27 above)
Comment (shamisp) [reply src/ucp/api/ucp.h:4528]: So the API is valid for all endpoint and if the socket was not connected using socket cm  UCS_ERR_NOT_CONNECTED is not returned. This has to be properly documented.

Comment (leibin2014) [reply src/ucp/api/ucp.h:4528]: @shamisp I guess there is some misunderstanding here. We always use socket address to connect. But the EP's status can have two cases:
1) If the EP is connected, then the API is valid to get local sockaddr and remote sockaddr as documented above, and return UCS_OK.
2) If the EP is not connected, then return UCS_ERR_NOT_CONNECTED as the error status.

I added this description according to @yosefe 's comment https://github.com/openucx/ucx/pull/7020#discussion_r664354439.
@yosefe Please correct me if I made any misunderstanding about the description.

Comment (yosefe) [reply src/ucp/api/ucp.h:4528]: there are cases when ep is connecting to worker address (not socket address) so there is no CM lane and we are returning UCS_ERR_NOT_CONNECTED. That's why we wrote "Valid only for endpoints created by connecting to a socket address." Perhaps should add "If this field is specified for an endpoint not connected to a socket address, the behavior is undefined" or ".. UCS_ERR_NOT_CONNECTED will be returned"

Comment (leibin2014) [reply src/ucp/api/ucp.h:4530]: @shamisp @yosefe Added. Thanks for the comments!

Comment (shamisp) [reply src/ucp/api/ucp.h:4530]: @yosefe "behavior is undefined" is the worse ...

Comment (leibin2014) [reply src/ucp/api/ucp.h:4530]: @shamisp  I added "If this field is specified for an endpoint not connected to a socket address, UCS_ERR_NOT_CONNECTED will be returned."

Comment (leibin2014) [reply src/ucp/api/ucp.h:4538]: @shamisp Could please approve this PR? Let me know please if you have any more comments. Thanks!

**Review by shamisp — APPROVED** [review]

**Comment by shamisp** [comment]
@tonycurtis can you please review documentation. thanks

**Review by tonycurtis — APPROVED** [review]

**Review by yosefe — APPROVED** [review]
pls squash

**Comment by leibin2014** [comment]
> pls squash

@yosefe Done

**Pull request merged**